Adaptive Tool README refinements (#2819)
microsoft/BotBuilder-Samples: experimental/adaptive-tool/README.md

# Bot Framework Adaptive Tool

> Bot Framework Adaptive Tool is a VS Code extension that helps developers handle LG (.lg), LU (.lu), and Dialog (.dialog) files efficiently. Adaptive Tool has tools and settings that make it easy to debug, analyze and enhance you language files.

Reviewer: >LU (.lu) [](start = 90, length = 8)

Doesn't it work on .qna as well?  (Which is basically lu format.)

Author: No, only .lu is currently supported:
![qna-lu-adaptive-tool](https://user-images.githubusercontent.com/9682548/96307883-2a6ab480-0fb7-11eb-9a05-9e1a675852d2.png)

Reviewer: Did you try adding qna as an lg extension?

---
In reply to: [506717510](https://github.com/microsoft/BotBuilder-Samples/pull/2819#discussion_r506717510) [](ancestors = 506717510)

Author: @chrimc62 No, I didn't. I just followed the instructions in the original README (install the extension, then open an lg, lu, or dialog file).

1. Open an LG, LU, or Dialog file and the extension will activate.

### Install from a local VSIX file

1. Clone the Adaptive Tool [repository](https://github.com/microsoft/BotBuilder-Samples/tree/main/experimental/adaptive-tool). In a terminal, navigate to the repository folder and run the following commands:

    ```cmd
    npm install
    npm run build
    ```

1. If `vsce` is not installed globally run `npm install -g vsce`.
1. Run `vsce package` to export the VSIX file. Now you're ready to install
1. Open Visual Studio Code and navigate to the **Extensions**.
1. Click the three dots in the top right corner of the Extensions. Then click **Install from VSIX...** and select the VSIX file you created earlier.
1. Open an LG, LU, or Dialog file and the extension will activate.

## Language features

Language features are driven by the [language server protocol](./languageServer.md). See the [Language Server Extension Guide](https://code.visualstudio.com/api/language-extensions/language-server-extension-guide) for more information.

### Syntax highlighting, diagnostic checks, and autocompletion

#### LU

> Different colors and styles for `intent`, `entity`, and `comment` components:

Author: I removed the formatting from both sections.

Formatting warnings and errors:
![lg_diagnostic](https://raw.githubusercontent.com/microsoft/BotBuilder-Samples/main/experimental/adaptive-tool/resources/images/lg_diagnostic.gif)

### Hover, suggestions, and navigation

#### LG

Template reference hover:
![template_hover](https://raw.githubusercontent.com/microsoft/BotBuilder-Samples/main/experimental/adaptive-tool/resources/images/template_hover.png)

Prebuilt function hover:
![function_hover](https://raw.githubusercontent.com/microsoft/BotBuilder-Samples/main/experimental/adaptive-tool/resources/images/function_hover.png)

Suggestions for templates and functions:
![function_template_suggestion](https://raw.githubusercontent.com/microsoft/BotBuilder-Samples/main/experimental/adaptive-tool/resources/images/function_template_suggestion.gif)

Structure property suggestions:
![structure_suggestion](https://raw.githubusercontent.com/microsoft/BotBuilder-Samples/main/experimental/adaptive-tool/resources/images/structure_suggestion.gif)

Template navigation:

![template_definition](https://raw.githubusercontent.com/microsoft/BotBuilder-Samples/main/experimental/adaptive-tool/resources/images/template_definition.gif)

## Debugging

> Adaptive Tool lets developers debug LG, LU, and Dialog files in runtime. This section covers the steps and shows and example of how to [configure](#configuration), [initialize](#start-bot-runtime-and-complete-initialization), and [debug](#debug-the-runtime) a bot.

Reviewer: >Adaptive Tool lets developers debug LG, LU, and Dialog files in runtime. [](start = 0, length = 72)

You are not describing the key benefit here.  "In the debugger you set breakpoints in your code, but that corresponds to classes and methods.  Setting breakpoints in dialogs, lg and lu allows you to get breakpoints in specific instances of that class rather than a code breakpoint which applies to all instances."

> Adaptive Tool lets developers debug LG, LU, and Dialog files in runtime. This section covers the steps and shows and example of how to [configure](#configuration), [initialize](#start-bot-runtime-and-complete-initialization), and [debug](#debug-the-runtime) a bot.

Reviewer 2: This is one reason I dislike the term _runtime_ for the bot executable. If they are talking about debugging a bot at run time, this edit is okay. If they're talking about debugging the executable at design time or run time, then...ask me and we can figure something out.
```suggestion
Adaptive Tool lets developers debug LG, LU, and dialog files at run time. This section covers the steps and shows an example of how to [configure](#configuration), [initialize](#start-bot-runtime-and-complete-initialization), and [debug](#debug-the-runtime) a bot.
```

### Configuration

- [Install](#getting-started) the Adaptive Tool extension.

> - Open the LG, LU ,or Dialog file to debug.

Reviewer 2: ```suggestion
- Open the LG, LU ,or dialog file to debug.
```

### Install from the Visual Studio Marketplace

1. Install the [Bot Framework Adaptive Tool](https://marketplace.visualstudio.com/items?itemName=adaptive-tool) from the [VS Code](https://marketplace.visualstudio.com/vscode) tab of the Visual Studio Marketplace.
1. Open an LG, LU, or dialog file and the extension will activate.

### Install from a local VSIX file

1. Clone the [BotBuilder-Samples](https://github.com/microsoft/BotBuilder-Samples) repo.
1. In a terminal or command prompt, navigate to the directory containing the Adaptive Tool, **experimental/adaptive-tool**, and run the following commands:

    ```cmd
    npm install
    ```
    
    ```cmd
    npm run build
    ```
    
  1. Run the command below to install [vsce](https://www.npmjs.com/package/vsce):
  
      ```cmd
      npm install -g vsce
      ```
    
1. Finally, run the following command to export the VSIX file:

    ```cmd
    vsce package
    ```
    
    You're now ready to install Adaptive Tool.
1. Open VS Code and open the **Extensions** view.

> 1. Click the **More Actions** button (...) in the upper-right corner of the window. Then click **Install from VSIX..** and select the VSIX file you created earlier.

Reviewer 2: I didn't notice the first time through that there are at least two different ellipsis buttons visible. Perhaps:
```suggestion
1. Click the **Views and More Actions** button (...) in the upper-right corner of the Extensions panel. Then click **Install from VSIX..** and select the VSIX file you created earlier.
```

}
```

`debugServer` refers to the port bot runs on. The default value is `4712`.

> ### Start bot runtime and complete initialization

Reviewer 2: Perhaps
```suggestion
### Start the bot and complete initialization
```
Links will need updating

> 1. Start a bot runtime. For this example we'll start the bot project `TodoBot` in [SampleBots](https://github.com/microsoft/botbuilder-dotnet/tree/main/tests/Microsoft.Bot.Builder.TestBot.Json).
> 1. Make sure the debugger port has been registered in `BotFrameworkHttpAdapter` with the `UseDebugger` method.
> 1. There are several ways to initialize a bot, and [BotFramework Emulator](https://github.com/microsoft/BotFramework-Emulator) is a typical approach. Open the Emulator and attach it to the bot to finish the initialization.

Reviewer 2: ```suggestion
1. Start a bot. For this example we'll start the bot project `TodoBot` in [SampleBots](https://github.com/microsoft/botbuilder-dotnet/tree/main/tests/Microsoft.Bot.Builder.TestBot.Json).
1. Make sure the debugger port has been registered in `BotFrameworkHttpAdapter` with the `UseDebugger` method.
1. There are several ways to initialize a bot, and the [BotFramework Emulator](https://github.com/microsoft/BotFramework-Emulator) is a typical approach. Open the Emulator and attach it to the bot to finish the initialization.
```

> ### Debug the runtime

Reviewer 2: ```suggestion
### Debug the bot
```
We're really not debugging the bot runtime/executable so much as we're debugging the resource files (.lu/.lg/.dialog).

> ### Debug the runtime

Reviewer: >the runtime [](start = 10, length = 11)

your bot

> 1. Run the Visual Studio Code program by clicking **F5** and set break points in the LG, LU, and/or Dialog files.

Reviewer 2: ```suggestion
1. Run the VS Code program by clicking **F5** and set break points in the LG, LU, and dialog files.
```

1. Chat with the bot in the Emulator.
1. If the extension is working properly the cursor will stop when the code hits any of corresponding breakpoints.

![breakpoints](../adaptive-tool/resources/images/breakpoints.gif)

## Adaptive Tool settings

Settings for LG files can be found under **LG** in the Extensions settings.

|Setting name|Description|
|-----|---------------|

> |`LG.Expression.ignoreUnknownFunction`|Show the diagnostic severity level of unknown functions in a LG file.<br><br>The levels include:<br/>`error`: treat unknown functions as error diagnostic<br>`warn`: treat unknown functions as warning diagnostic<br>`ignore`: ignore unknown functions|
> |`LG.Expression.customFunctionList`| Create a comma-separated customized function list (example: a, b, c). You can use both custom functions added in your logic as well as functions added to the `customFunctionList` setting.|

Reviewer 2: ```suggestion
|`LG.Expression.ignoreUnknownFunction`|Set the severity diagnostic level for unknown functions in a LG file.<br><br>The levels include:<br/>`error`: treat unknown functions as an error<br>`warn`: treat unknown functions as a warning<br>`ignore`: ignore unknown functions|
|`LG.Expression.customFunctionList`| Create a comma-separated customized function list (example: a, b, c). You can use both custom functions added in your logic as well as functions added to the `customFunctionList` setting.|
```
